### PR TITLE
Crangler

### PR DESCRIPTION
**Pull request by kroening**
This adds a new command-line tool, which replaces current flows that enact changes to the program that is analysed. The new flow is to make these changes before parsing, which enables more far-reaching changes and changes that depend on scoped identifiers.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
Two high-level comments:
1) Do we really need YAML? Couldn't we use JSON instead and have people use tools like `yq` to convert?
2) Even if not a particularly elaborate regression test, could we at least have an example of an application?

File: src/crangler/crangler_parse_options.cpp
Comment (tautschnig) [inline src/crangler/crangler_parse_options.cpp:28]: Use error logging via the message handler instead?

File: src/crangler/Makefile
Comment (tautschnig) [inline src/crangler/Makefile:1]: Somewhat arbitrary place for comment: we're missing a CMake configuration for this new tool.

File: src/crangler/c_defines.cpp
Comment (tautschnig) [inline src/crangler/c_defines.cpp:37]: Do we need to care about `#undef` and a possible re-definition?

File: src/crangler/c_defines.cpp
Comment (tautschnig) [inline src/crangler/c_defines.cpp:64]: What about macros that use other macros? What about macros that take parameters (see `getc_unlocked(fp)` example in the comment)?

File: src/crangler/c_defines.h
Comment (tautschnig) [inline src/crangler/c_defines.h:28]: Never used (although they'd need to be taken into account).

File: src/crangler/c_defines.h
Comment (tautschnig) [inline src/crangler/c_defines.h:23]: A brief documentation of the purpose of this class would be useful.

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: There aren't a lot of comments in this code, but then this one stating the obvious. Perhaps I'm missing something and it's not actually doing the obvious?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: See above.

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Use `join_strings`?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp]: Perhaps "output" instead of "outputs?"

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp:340]: Should we have a Windows/MSVC version thereof?

File: src/crangler/c_wrangler.cpp
Comment (tautschnig) [inline src/crangler/c_wrangler.cpp:362]: Windows/MSVC?

File: src/crangler/c_wrangler.h
Comment (tautschnig) [inline src/crangler/c_wrangler.h]: `class jsont;` might do?

File: src/crangler/mini_c_parser.cpp
Comment (tautschnig) [inline src/crangler/mini_c_parser.cpp]: Remove?

**Review by tautschnig — APPROVED** [review]
Approving after some further out-of-band discussion. In addition to my earlier comments: some further tests and/or documentation would be really helpful.

File: src/crangler/c_defines.h
Comment (kroening) [reply src/crangler/c_defines.h:23]: added a comment

File: src/crangler/c_wrangler.cpp
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.cpp
Comment (kroening) [reply src/crangler/c_wrangler.cpp]: done

File: src/crangler/c_wrangler.h
Comment (kroening) [reply src/crangler/c_wrangler.h]: done

File: src/crangler/mini_c_parser.cpp
Comment (kroening) [reply src/crangler/mini_c_parser.cpp]: done

**Comment by tautschnig** [comment]
We might still want to add a capability to perform name mangling, but once done we should be able to get rid of the `export-file-local-symbols` cross-cutting changes.

**Review by TGWDB — APPROVED** [review]
This is an approval for the Cmake/make aspects as these were blocked by OpenSource/Codeowner

**Pull request merged**
